### PR TITLE
Fix bug in parsing empty birthday from storage

### LINKED ISSUE #105: 🐞 Empty birthday is treated as illegal value and causes data file to be read as incorrect format
### Describe the issue

The Birthday field should be optional when adding new contacts, e.g. add -n John but in the current implementation, when it is saved in the data file as "", it is interpreted as incorrect format as seen in the below screenshot.

image

Expected outcome: Loads the data file like normal, but instead the data file is interpreted as invalid and starts with an empty addressbook.

Steps to reproduce:
1. Delete addressbook.json in data (to start with prepopulated sample data)
2. Add a contact without the birthday field: add -n John
3. Close the application
4. Reopen the application. Upon attempting to load data, the data file will be interpreted as invalid and the above logs will be shown.

________________________________________________________________
### Additional context

________________________________________________________________
### Environment
 - OS: Windows 10
 - JDK version: 11

### PR DESCRIPTION
**Pull request by zhengruoxin**
Fixes #105

### CODE REVIEW

File: src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java [hunk 1]
```diff
@@ -115,12 +115,14 @@ public Person toModelType() throws IllegalValueException {
...
-            throw new IllegalValueException((Birthday.MESSAGE_YEAR_CONSTRAINTS));
-        } catch (IllegalArgumentException err) { // date in wrong format
-            throw new IllegalValueException(Birthday.MESSAGE_CONSTRAINTS);
+        else {
+            try {
+                modelBirthday = new Birthday(birthday);
+            } catch (DateTimeException err) { // birthday year exceeds current year
+                throw new IllegalValueException((Birthday.MESSAGE_YEAR_CONSTRAINTS));
+            } catch (IllegalArgumentException err) { // date in wrong format
+                throw new IllegalValueException(Birthday.MESSAGE_CONSTRAINTS);
```
Comment (garyljj) [inline src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java:124]: Swap error message
```suggestion
            } catch (DateTimeException err) { // date in wrong format
                throw new IllegalValueException((Birthday.MESSAGE_CONSTRAINTS));
            } catch (IllegalArgumentException err) { // birthday year exceeds current year
                throw new IllegalValueException(Birthday.MESSAGE_YEAR_CONSTRAINTS);
```

**Review by garyljj — CHANGES_REQUESTED** [review]
Error message swapped.

Also I noticed, in Birthday.java, suggestion to switch the order of checks
check `isValidBirthday()` before `isValidBirthdayYear()`
Cos if invalid birthday, eg "abc-01-01"
`isValidBirthdayYear()` would actually throw a `NumberFormatException`, which just so happen to be caught by `IllegalArgumentException` as it is a subclass.
But should actually show invalid birthday format error

File: src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java [hunk 2]
```diff
@@ -115,12 +115,14 @@ public Person toModelType() throws IllegalValueException {
         if (birthday.equals(Birthday.EMPTY_BIRTHDAY_STRING)) {
             modelBirthday = Birthday.EMPTY_BIRTHDAY;
         }
-        try {
-            modelBirthday = new Birthday(birthday);
-        } catch (DateTimeException err) { // birthday year exceeds current year
-            throw new IllegalValueException((Birthday.MESSAGE_YEAR_CONSTRAINTS));
-        } catch (IllegalArgumentException err) { // date in wrong format
-            throw new IllegalValueException(Birthday.MESSAGE_CONSTRAINTS);
+        else {
```
Comment (pyuxiang) [inline src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java:118]: Might want to change to `} else {` style recommended in this module.

**Review by pyuxiang — DISMISSED** [review]
Other than error message issue mentioned by @garyljj, LGTM.

File: src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java (hunk 1 above)
Comment (pyuxiang) [reply src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java:124]: Screenshot of what @garyljj is referring to:

image

File: src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java [hunk 3]
```diff
@@ -114,13 +114,14 @@ public Person toModelType() throws IllegalValueException {
...
-            modelBirthday = new Birthday(birthday);
-        } catch (DateTimeException err) { // birthday year exceeds current year
-            throw new IllegalValueException((Birthday.MESSAGE_YEAR_CONSTRAINTS));
-        } catch (IllegalArgumentException err) { // date in wrong format
-            throw new IllegalValueException(Birthday.MESSAGE_CONSTRAINTS);
+        } else {
+            try {
+                modelBirthday = new Birthday(birthday);
+            } catch (DateTimeException err) { // date in wrong format
+                throw new DateTimeException(Birthday.MESSAGE_CONSTRAINTS);
```
Comment (garyljj) [inline src/main/java/seedu/partyplanet/storage/JsonAdaptedPerson.java:121]: IllegalValueException was correct haha
IllegalValueException is the one caught by storage. Or else an invalid date in the save file will not be handled and program will terminate

```suggestion
                throw new IllegalValueException(Birthday.MESSAGE_CONSTRAINTS);
```

image

**Review by garyljj — CHANGES_REQUESTED** [review]
Minor mistake in exception thrown

**Review by pyuxiang — APPROVED** [review]
LGTM

**Review by garyljj — APPROVED** [review]
LGTM

**Pull request merged**